File: Prision/src/main/java/com/gkzxhn/prison/service/EReportService.java
package com.gkzxhn.prison.service;

import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.gkzxhn.prison.async.SingleRequestQueue;
import com.gkzxhn.prison.async.VolleyUtils;
import com.gkzxhn.prison.common.Constants;
import com.gkzxhn.prison.utils.XtHttpUtil;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;


/**
 * 负责接收异步事件并通过广播发送的服务
 * Created by lv on 2017/1/11.
 */

public class EReportService extends Service {

    private static final String TAG = "EReportService";

    private String qid;

    private VolleyUtils volleyUtils=new VolleyUtils();
    @Override
    public void onCreate() {
        super.onCreate();
        ResetQuest();
    }

    //重置事件队列id
    public void ResetQuest() {
        try {
            volleyUtils.get(JSONObject.class, XtHttpUtil.RESET, null, new VolleyUtils.OnFinishedListener<JSONObject>() {
                @Override
                public void onSuccess(JSONObject response) {
                    Log.d(TAG, "Reset" + response.toString());
                    try {
                        JSONObject object = response.getJSONObject("v");
                        qid = object.getString("qid");
                        int next_seq = object.getInt("next_seq");//如果无效qid
                        GetClearQuest(next_seq);
                        GetQueryQuest(next_seq + 1);
                    } catch (JSONException e) {
//                            e.printStackTrace();
                    }
                }

                @Override
                public void onFailed(VolleyError error) {
                    ResetQuest();//如果访问拒绝，继续访问
                    Log.d(TAG, "ResetQuest..." + error.toString());
                }
            });
        } catch (AuthFailureError authFailureError) {
            authFailureError.printStackTrace();
        }
    }


    //清除队列
    private void GetClearQuest(final int next_seq) {
        String url=XtHttpUtil.CLEAR + qid + "&expect_seq=" + next_seq;
        try {
            volleyUtils.get(JSONObject.class, url, null, new VolleyUtils.OnFinishedListener<JSONObject>() {
                @Override
                public void onSuccess(JSONObject response) {

                    //Log.d(TAG, "Clear"+response.toString());

//                        init(response);
                }

                @Override
                public void onFailed(VolleyError error) {
                    //                Log.d(TAG, error.toString() + "...");
                }
            });
        } catch (AuthFailureError authFailureError) {
            authFailureError.printStackTrace();
        }
    }

//    private void init(JSONObject response) {
//        try {
//            JSONObject obj = response.getJSONObject("v");
//            String e = response.getString("e");
//            int next_seq = obj.getInt("next_seq");
////            GetQueryQuest(next_seq + 1);
////            if (e.equals("cleared")) {
////                GetQueryQuest(next_seq + 1);
////            }
//            if (e.equals("event_lost")) {
//                GetQueryQuest(next_seq);
//            } else if (e.equals("cleared")) {
//                GetQueryQuest(next_seq + 1);
//            }
//        } catch (JSONException e) {
////            e.printStackTrace();
//        }
//    }



    //查询事件
    public void GetQueryQuest(final int a) {
        String url=XtHttpUtil.QUERY + qid + "&expect_seq=" + (a);
        try {
            volleyUtils.get(JSONObject.class, url, null, new VolleyUtils.OnFinishedListener<JSONObject>() {
                @Override
                public void onSuccess(JSONObject response) {
                    initData(response, a);
                }

                @Override
                public void onFailed(VolleyError error) {
                    GetQueryQuest(a);
                    Log.d(TAG, "GetQueryQuest: " + error.toString());
                }
            });
        } catch (AuthFailureError authFailureError) {
            authFailureError.printStackTrace();
        }
    }

    //{"service":"EventQueue","seq":0,"v":{"expect_seq":106,"next_seq":105},"e":"event_lost"}
    private void initData(JSONObject response, int a) {
        try {
            if (response.getString("e").equals("event_lost")) {
                GetQueryQuest(response.getJSONObject("v").getInt("next_seq"));
            } else {
                if (response.getString("e").equals("qid_invalid")) {
                    ResetQuest();
                } else {
                    sendLocalBroad(response, Constants.ZIJING_ACTION);
                }
                GetQueryQuest(a + 1);
            }
            Log.i(TAG, "initData: response >>> " + response.toString());

        } catch (JSONException e) {
            e.printStackTrace();
        }

    }


    //发送本地广播
    private void sendLocalBroad(JSONObject response, String action) {

        Intent intent = new Intent();
        intent.setAction(action);
        intent.putExtra(Constants.ZIJING_JSON, response.toString());
        sendBroadcast(intent);
//        Log.d(TAG,response.toString());
    }





    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }



    @Override
    public void onDestroy() {
        super.onDestroy();
        stopSelf();
    }

}

File: Prision/src/main/java/com/gkzxhn/prison/service/EReportService.kt
package com.gkzxhn.prison.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log

import com.android.volley.AuthFailureError
import com.android.volley.VolleyError
import com.gkzxhn.prison.common.Constants
import com.gkzxhn.prison.utils.XtHttpUtil
import com.gkzxhn.wisdom.async.SingleRequestQueue
import com.gkzxhn.wisdom.async.VolleyUtils

import org.json.JSONException
import org.json.JSONObject


/**
 * 负责接收异步事件并通过广播发送的服务
 * Created by lv on 2017/1/11.
 */

class EReportService : Service() {
    private val TAG = EReportService::class.java.name
    private var qid: String? = null
    private val volleyUtils = VolleyUtils()
    override fun onCreate() {
        super.onCreate()
        //取消所有请求
        SingleRequestQueue.instance.cancelAll(TAG)
        ResetQuest()
    }

    //重置事件队列id
    fun ResetQuest() {
        try {
            volleyUtils.get(JSONObject::class.java, XtHttpUtil.RESET, TAG, object : VolleyUtils.OnFinishedListener<JSONObject> {
                override fun onSuccess(response: JSONObject) {
                    Log.d(TAG, "Reset" + response.toString())
                    try {
                        val `object` = response.getJSONObject("v")
                        qid = `object`.getString("qid")
                        val next_seq = `object`.getInt("next_seq")//如果无效qid
                        GetClearQuest(next_seq)
                        GetQueryQuest(next_seq + 1)
                    } catch (e: JSONException) {
                        //                            e.printStackTrace();
                    }

                }

                override fun onFailed(error: VolleyError) {
                    ResetQuest()//如果访问拒绝，继续访问
                    Log.d(TAG, "ResetQuest..." + error.toString())
                }
            })
        } catch (authFailureError: AuthFailureError) {
            authFailureError.printStackTrace()
        }

    }


    //清除队列
    private fun GetClearQuest(next_seq: Int) {
        val url = XtHttpUtil.CLEAR + qid + "&expect_seq=" + next_seq
        try {
            volleyUtils.get(JSONObject::class.java, url, TAG, object : VolleyUtils.OnFinishedListener<JSONObject> {
                override fun onSuccess(response: JSONObject) {

                    //Log.d(TAG, "Clear"+response.toString());

                    //                        init(response);
                }

                override fun onFailed(error: VolleyError) {
                    //                Log.d(TAG, error.toString() + "...");
                }
            })
        } catch (authFailureError: AuthFailureError) {
            authFailureError.printStackTrace()
        }

    }

    //    private void init(JSONObject response) {
    //        try {
    //            JSONObject obj = response.getJSONObject("v");
    //            String e = response.getString("e");
    //            int next_seq = obj.getInt("next_seq");
    ////            GetQueryQuest(next_seq + 1);
    ////            if (e.equals("cleared")) {
    ////                GetQueryQuest(next_seq + 1);
    ////            }
    //            if (e.equals("event_lost")) {
    //                GetQueryQuest(next_seq);
    //            } else if (e.equals("cleared")) {
    //                GetQueryQuest(next_seq + 1);
    //            }
    //        } catch (JSONException e) {
    ////            e.printStackTrace();
    //        }
    //    }


    //查询事件
    fun GetQueryQuest(a: Int) {
        val url = XtHttpUtil.QUERY + qid + "&expect_seq=" + a
        try {
            volleyUtils.get(JSONObject::class.java, url, TAG, object : VolleyUtils.OnFinishedListener<JSONObject> {
                override fun onSuccess(response: JSONObject) {
                    initData(response, a)
                }

                override fun onFailed(error: VolleyError) {
                    GetQueryQuest(a)
                    Log.d(TAG, "GetQueryQuest: " + error.toString())
                }
            })
        } catch (authFailureError: AuthFailureError) {
            authFailureError.printStackTrace()
        }

    }

    //{"service":"EventQueue","seq":0,"v":{"expect_seq":106,"next_seq":105},"e":"event_lost"}
    private fun initData(response: JSONObject, a: Int) {
        try {
            if (response.getString("e") == "event_lost") {
                GetQueryQuest(response.getJSONObject("v").getInt("next_seq"))
            } else {
                if (response.getString("e") == "qid_invalid") {
                    ResetQuest()
                } else {
                    sendLocalBroad(response, Constants.ZIJING_ACTION)
                }
                GetQueryQuest(a + 1)
            }
            Log.i(TAG, "initData: response >>> " + response.toString())

        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }


    //发送本地广播
    private fun sendLocalBroad(response: JSONObject, action: String) {

        val intent = Intent()
        intent.action = action
        intent.putExtra(Constants.ZIJING_JSON, response.toString())
        sendBroadcast(intent)
        //        Log.d(TAG,response.toString());
    }


    override fun onBind(intent: Intent): IBinder? {
        return null
    }


    override fun onDestroy() {
        super.onDestroy()
        stopSelf()
    }
}


